### ISSUE TITLE
npm install ffi failing on Windows

### ISSUE THREAD

**Issue by pita2230** [node-ffi/node-ffi#280]
C:\Program Files\nodejs\node_modules\npm>npm install ffi

> ref@1.3.2 install C:\Program Files\nodejs\node_modules\npm\node_modules\ref
> node-gyp rebuild

C:\Program Files\nodejs\node_modules\npm\node_modules\ref>if not defined npm_config_node_gyp (node "C:\Program Files\nodejs\node_modules\npm\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node  rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
C:\Program Files\nodejs\node_modules\npm\node_modules\ref\build\binding.vcxproj(20,3): error MSB4019: The imported project "C:\Microsoft.Cpp.Default.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
gyp ERR! build error
gyp ERR! stack Error: `msbuild` failed with exit code: 1
gyp ERR! stack     at ChildProcess.onExit (C:\Program Files\nodejs\node_modules\npm\node_modules\node-gyp\lib\build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:100:13)
gyp ERR! stack     at ChildProcess.emit (events.js:185:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Windows_NT 6.1.7601
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Program Files\nodejs\node_modules\npm\node_modules\ref
gyp ERR! node -v v5.5.0
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok
npm WARN install:ref@1.3.2 ref@1.3.2 install: `node-gyp rebuild`
npm WARN install:ref@1.3.2 Exit status # 1****

**Comment by TooTallNate** [comment]
Do you have the build tools installed?

**Comment by pita2230** [comment]
I do. I followed all the below steps. Am I missing anything here?

$ npm install -g node-gyp
Python (v2.7.10 recommended, v3.x.x is not supported)
Make sure that you have a PYTHON environment variable, and it is set to drive:\path\to\python.exe not to a folder
Microsoft Visual Studio C++ 2013 (Express version works well)
For 64-bit builds of node and native modules you will also need the Windows 7 64-bit SDK
$ node-gyp --python /path/to/python2.7
$ npm config set python /path/to/executable/python2.7

Thanks!

**Comment by TooTallNate** [comment]
Definitely something setup related still. Try checking StackOverflow, google, etc. http://stackoverflow.com/q/16092169/376773

**Comment by hdf** [comment]
Just had the same problem. Python installation was the cause. 64 bit node needs 64 bit Python. Make sure to set PYTHONHOME and PYTHONPATH environment variables to the 64 bit version of your Python installation. Also, with `npm config set` the path has to include the exe file itself, not just the directory.

**Comment by Ciantic** [comment]
I'm getting a same problem. Node-gyp installs just fine but ffi does not want to get installed:

In administrative command prompt:
```
C:\WINDOWS\system32>npm install ffi -g

> ref@1.3.2 install C:\Copies\Nodist\bin\node_modules\ffi\node_modules\ref
> node-gyp rebuild


C:\Copies\Nodist\bin\node_modules\ffi\node_modules\ref>if not defined npm_config_node_gyp (node "C:\Copies\Nodist\npmv\3.10.8\bin\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "" rebuild )
Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
C:\Copies\Nodist\bin\node_modules\ffi\node_modules\ref\build\binding.vcxproj(20,3): error MSB4019: The imported project
 "C:\Microsoft.Cpp.Default.props" was not found. Confirm that the path in the <Import> declaration is correct, and that
 the file exists on disk.
```

**Comment by kroggen** [comment]
+1 for prebuilt binaries for Windows

**Comment by mStirner** [comment]
+1 same as #Ciantic

**Comment by keyiis** [comment]
+1 for prebuilt binaries for Windows10
```
D:\work\nodejs\keypos\test-dongle>npm config ls
; cli configs
metrics-registry = "https://registry.npmjs.com/"
scope = ""
user-agent = "npm/5.5.1 node/v8.9.3 win32 x64"

; userconfig C:\Users\Administrator\.npmrc
cache = "D:\\work\\nodejs\\global_modules\\cache"
msvs_version = "2015"
prefix = "D:\\work\\nodejs\\global_modules\\modules"
python = "C:\\Python27"
registry = "https://registry.npmjs.com/"

; globalconfig D:\work\nodejs\global_modules\modules\etc\npmrc
cache = "D:\\work\\nodejs\\global_modules\\cache"
msvs_version = "2015"
prefix = "D:\\work\\nodejs\\global_modules\\modules"

; builtin config undefined

; node bin location = C:\Program Files\nodejs\node.exe
; cwd = D:\work\nodejs\demo\test
; HOME = C:\Users\Administrator
; "npm config ls -l" to show all defaults.
```

**Comment by TwoAbove** [comment]
For future searchers:
I had an issue with `npm i --save ffi`  failing with 
```
E:\...\node_modules\ref\build\binding.vcxproj(20,3): error MSB4019: The imported project "E:\Microsoft.Cpp.Default.props" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
```
going through different solutions, I checked `npm config ls`, which gave me this line:
```
...
msvs_version = "2015"
...
```
I had visual studio 2017 installed. 
Changing it to `msvs_version = "2017"` (with `npm config set msvs_version 2017`) solved the issue!

TL;DR:
Check msvs_version with `npm config ls`. It may be different from the installed version

**Comment by Pointotech** [comment]
Still fails on Windows 10 with VS 2019. Why do I have to debug MSBuild issues when I'm using Node? Pretty confusing.
